### PR TITLE
feat: implement version prefix validation

### PR DESCRIPTION
**Pull request by fsrv-xyz**
add implementation for https://github.com/thenativeweb/get-next-version/issues/70

Signed-off-by: Florian Bauer <florian@fsrv.xyz>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -33,7 +33,7 @@ Optionally, you may hand over the `--repository` (or short `-r`) flag to specify
 $ get-next-version --repository <PATH>
 ```
 
-If you need to prefix the version, you can use the `--prefix` (or short `-p`) flag.
+If you need to prefix the version, you can use the `--prefix` (or short `-p`) flag. Note that the prefix must be a valid tag name on its own (see [implementation comments](git/tags.go#L65) for details).
```
Comment (goloroden) [inline README.md:36]: While the idea of having a reference to the file (and even the line) is basically great, it doesn't work out in the long run. The next commit (which maybe is completely unrelated to this one) may change the line number, or even the file's name or path – and I almost certainly can guarantee that whoever will do this change won't think of adjusting this comment in the `README.md` file.

So, while I like the sentence you added, I'd suggest to remove the part within the parentheses.

File: git/tags.go [hunk 2]
```diff
@@ -52,3 +53,25 @@ func GetAllSemVerTags(repository *git.Repository) (Tags, error) {
...
+
+/*
+IsValidTagName runs a series of regex checks to ensure that the tag name is valid
+  Tags cannot begin or end with, or contain multiple consecutive / characters.
+  They cannot contain any of the following characters \, ?, ~, ^, :, * , [, @.
+  They cannot contain a space.
+  They cannot end with a . or have two consecutive .. anywhere within them.
+  Tags are not case-sensitive.
+*/
+func IsValidTagName(tag string) bool {
```
Comment (goloroden) [inline git/tags.go:65]: I like that you put this validation into a file of its own, and that you even took care of adding a comment (which we are currently missing for all the other functions, but it's good that you started) 😊

However, there is one thing I'd ask you to change, and this is not to use negative validation, but positive: Do not check for characters that are forbidden, but check for characters that are allowed. Some weird unicode characters are not caught by the regexs below, but they could cause problems – if not with Git, then maybe with other tools.

So I'd suggest to use an allow list instead of a deny list, and – maybe for the start – simply allow `a-z`, `A-Z`, `0-9`, `_`, and `-` (as also described in the linked issue). In addition, we might also support the forward slash (`/`).

This way the check becomes way easier, and you only have to use one regular expression, not multiple ones.

Once this first check has been done, I'd suggest a second check to verify that some characteristics are met, such as "can not end with a number" (which, for obvious reasons, doesn't make sense), or "there shall be no two slashes following each other".

What do you think?

File: cli/root.go [hunk 3]
```diff
@@ -34,6 +35,10 @@ var RootCommand = &cobra.Command{
 			"version",
 		}
 
+		if !git.IsValidTagName(strings.TrimSuffix(rootPrefixFlag, ".")) {
```
Comment (goloroden) [inline cli/root.go:38]: Shouldn't the trailing dot be handled by the `IsValidTagName` function, so that the call to `TrimSuffix` can be removed here?

File: git/tags_test.go [hunk 4]
```diff
@@ -87,3 +87,32 @@ func TestGetAllSemVerTags(t *testing.T) {
...
+			assert.Equal(t, tt.expected, git.IsValidTagName(tt.tagName))
+		})
+	}
+}
+
+func BenchmarkIsValidTagName(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		git.IsValidTagName("v1.2.3")
+	}
+}
```
Comment (goloroden) [inline git/tags_test.go:118]: While I very much appreciate that you think about benchmark tests, I'm not sure if it's actually relevant (i.e. helpful) in the long run – but it is code that may need to be adjusted and maintained.

So, would you mind to remove this one?

**Review by goloroden — CHANGES_REQUESTED** [review]
Hey @fsrv-xyz 👋 

Thanks for bringing up another PR 🎉
That's great, thanks a lot 😊

As with the last time, I've left some comments.
Could you please review them and adjust the code as needed?

File: cli/root.go (hunk 3 above)
Comment (fsrv-xyz) [reply cli/root.go:38]: In this case, I try to address a case where the prefix should be something like `usa.` (resulting in `usa.1.2.3`). The prefix on its own would not be a valid tag name. Due to the dot is not at the end of the version, the resulting one should be valid.

While I am writing this comment, I see that solving that use case is not even as simple as I thought (trailing slashes/dots, ...).

Will be rewritten soon.

File: git/tags_test.go (hunk 4 above)
Comment (fsrv-xyz) [reply git/tags_test.go:118]: Due to the benchmark is only running on its own, it would only show massive performance decrease by added functionality.

Can actually be removed because the implementation should not be frequently changed in the future.

File: README.md (hunk 1 above)
Comment (fsrv-xyz) [reply README.md:36]: I see the problem you are addressing. It kicks me out of my perfect world where everyone is using the refactor function of all those wonderful Czech IDEs.

**Comment by fsrv-xyz** [comment]
If you don't mind, I'd like to sort the variables using the tool [goimportssort](https://github.com/bonsai-oss/goimportssort) which we forked and modified a few weeks ago.

**Comment by goloroden** [comment]
@fsrv-xyz Regarding the `goimportssort` tool – please don't make yourself too much effort here, since the final decision on sorting will be done by the IDE anyway (in our case, VSCode with the Go plugin), so it may be, that depending on the sort order, it gets overwritten by the next change anyway.

File: cli/root.go [hunk 5]
```diff
@@ -34,6 +34,10 @@ var RootCommand = &cobra.Command{
 			"version",
 		}
 
+		if valid, prefixValidationError := util.IsValidTagPrefix(rootPrefixFlag); !valid {
```
Comment (goloroden) [inline cli/root.go:37]: Just a small change here, please:

```suggestion
		if isValid, prefixValidationError := util.IsValidTagPrefix(rootPrefixFlag); !isValid {
```

File: util/prefix_validate.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package util
```
Comment (goloroden) [inline util/prefix_validate.go:1]: I understand why you moved the file to the `util` package, and I'm basically fine with this.

However, I think we should adjust a few things:

1. The function is still called `IsValidTagPrefix`, although it is not related directly to Git tags any more. Shouldn't it be `IsValidVersionPrefix` then?
2. If we change the name like this, I'd suggest to name the file `version_prefix.go`, since it then contains all the helper methods for version prefixes.

File: util/prefix_validate.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+	"regexp"
+)
+
+/*
+IsValidTagPrefix runs a series of regex checks to ensure that the tag prefix is valid
+  Tags cannot begin or end with, or contain multiple consecutive / characters.
+  They cannot contain any of the following characters \, ?, ~, ^, :, * , [, @.
+  They cannot contain a space.
+  They cannot end with a . or have two consecutive .. anywhere within them.
+  Tags are not case-sensitive.
```
Comment (goloroden) [inline util/prefix_validate.go:14]: This description does not match exactly what the function does any more. I assume it's better to also describe what is allowed here then to describe what's forbidden.

File: util/prefix_validate.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+	}
+	if regexp.MustCompile(`/{2,}`).MatchString(prefix) {
+		return false, errors.New("tag prefix contains multiple consecutive / characters")
+	}
+	if regexp.MustCompile(`\.{2,}`).MatchString(prefix) {
+		return false, errors.New("tag prefix contains multiple consecutive . characters")
+	}
+	if !regexp.MustCompile(`\D$`).MatchString(prefix) {
+		return false, errors.New("tag prefix ends with a digit")
+	}
```
Comment (goloroden) [inline util/prefix_validate.go:32]: Can you please move the regexs to package level variables, and call `MustCompile` in the `init` function? For now, it doesn't make a difference, but if we ever call this function twice, we save time on executing the regexs, as we don't have to recompile them over and over again.

Besides with (this is something you could not know, sorry for that): Can you rephrase the error messages to use the `must` form, e.g. `version prefix must not contain …`?

Oh, and it should be `version prefix`, not `tag prefix` 😉

File: util/prefix_validate_test.go [hunk 9]
```diff
@@ -0,0 +1,40 @@
+package util_test
```
Comment (goloroden) [inline util/prefix_validate_test.go:1]: When we change the file name of the `prefix_validate.go` file, we need to think of renaming this file as well.

File: util/prefix_validate_test.go [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/thenativeweb/get-next-version/util"
+)
+
+func TestIsValidTagPrefix(t *testing.T) {
```
Comment (goloroden) [inline util/prefix_validate_test.go:10]: … and this function 😉

File: util/prefix_validate_test.go [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+	for _, tt := range []struct {
```
Comment (goloroden) [inline util/prefix_validate_test.go:11]: Can we find a better (i.e. more descriptive name for `tt`)?

(Actually, right now, I couldn't say what `tt` stands for. What about e.g. `testcase`?)

**Review by goloroden — CHANGES_REQUESTED** [review]
Hi @fsrv-xyz 👋 

Thanks for rewriting this PR 😊 
Again, I've left a few comments, which I would like to ask that you have a look at.

Besides, a general note, because I've now seen some code you've written: My general impression of your code is pretty good. There has not been any typo so far (e.g. in error messages), it's formatted in a reasonable way (blank lines to build blocks), … and I just wanted to say that I like it, and that I appreciate what you do. Thanks a lot 😊

**Comment by fsrv-xyz** [comment]
@goloroden Thank you for your compliments. 👍 
It's just nice when you can still learn something even after years in software development and see the perspective of other developers on your work.
I think, there are only a few communities with such a fantastic experience. 😊

For me, working on the tools we/I use is a great honor to me and a good chance to give something back to the community.

However, I have applied the changes you suggested as well as a few style/test fixes.

**Review by goloroden — APPROVED** [review]
